cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/AbstractKafkaSampleStore.java

```
@@ -63,23 +63,43 @@ protected void createProducer(Map<String, ?> config, String producerClientId) {
-      } catch (InterruptedException | ExecutionException | TimeoutException e) {
-        throw new IllegalStateException("Auto creation of sample store topics failed due to failure to describe cluster.", e);
-      }
-      if (numberOfBrokersInCluster <= 1) {
-        throw new IllegalStateException(String.format("Kafka cluster has less than 2 brokers (brokers in cluster=%d, zookeeper.connect=%s)",
-                                                      numberOfBrokersInCluster, config.get(ExecutorConfig.ZOOKEEPER_CONNECT_CONFIG)));
+        try {
+          numberOfBrokersInCluster = (short) adminClient.describeCluster().nodes().get(CLIENT_REQUEST_TIMEOUT_MS,
+                  TimeUnit.MILLISECONDS).size();
+        } catch (InterruptedException | ExecutionException | TimeoutException e) {
+          throw new IllegalStateException("Auto creation of sample store topics failed due to failure to describe cluster.", e);
+        }
+        if (numberOfBrokersInCluster <= 1) {
+          throw new IllegalStateException(String.format("Kafka cluster has less than 2 brokers (brokers in cluster=%d, zookeeper.connect=%s)",
+                  numberOfBrokersInCluster, config.get(ExecutorConfig.ZOOKEEPER_CONNECT_CONFIG)));
+        }
+
+        return (short) Math.min(DEFAULT_SAMPLE_STORE_TOPIC_REPLICATION_FACTOR, numberOfBrokersInCluster);
+      } catch (IllegalStateException e) {
+        if (retryCount + 1 == maxRetryCount) {
+          throw e;
+        }
       }
 
-      return (short) Math.min(DEFAULT_SAMPLE_STORE_TOPIC_REPLICATION_FACTOR, numberOfBrokersInCluster);
+      retryCount++;
+      try {
+        TimeUnit.MILLISECONDS.sleep(waitingTime);
+      } catch (InterruptedException ignored) {
+      }
     }
 
-    return _sampleStoreTopicReplicationFactor;
+    return (short) Math.min(DEFAULT_SAMPLE_STORE_TOPIC_REPLICATION_FACTOR, numberOfBrokersInCluster);
```

Would the code execution ever reach this line? If not, we should just throw an exception here.

---

No, this line is unreachable. OK, I will throw an exception here.